beaker_kernel/lib/agent_tasks.py

```diff
@@ -0,0 +1,91 @@
+import asyncio
+import json
+
+from archytas.react import ReActAgent
+
+OUTPUT_CHAR_LIMIT = 1000
+
+class Summarizer(ReActAgent):
+
+    def __init__(
+        self,
+        notebook: dict = {},
+        **kwargs,
+    ):
+        super().__init__(
+            model="gpt-4-turbo-preview",
+            # api_key=api_key,  # TODO: get this from configuration
+            tools=[],
+            verbose=False,
+            spinner=None,
+            rich_print=False,
+            allow_ask_user=False,
+            **kwargs
+        )
+
+        context = f"""
+You are working to provide summaries of this provided LLM-Powered Beaker notebook:
+```
+{notebook}
+```
+
+Beaker notebooks are an extension of Jupyter Notebooks that supports a new subtype of
+cell called a 'query' cell. This means that some markdown cells with the `beaker_cell_type`
+of `query` contain the the user query, the agents thoughs and possibly the final response
+of the agent. If a code cell has a parent_id, that means it was produced by the LLM Agent
+in the parent query cell.
+"""
+        self.add_context(context)
+        
+
+async def summarize(notebook: dict,
+    summary_types: tuple[str, ...] = (
+        "a single sentence BLUF that must be in past tense",
+        "a summary",
+    )
+):
+    for cell in notebook["cells"]:
+        if "outputs" in cell and len(cell["outputs"]) > 0:
+            for output in cell["outputs"]:
+                if "text" in output:
+                    output["text"] = output["text"][:OUTPUT_CHAR_LIMIT]
+                if "traceback" in output:
+                    del output["traceback"]
+                if "data" in output:
+                    for data_type in output["data"]:
+                        output["data"][data_type] = output["data"][data_type][:OUTPUT_CHAR_LIMIT]
+            
+    agent = Summarizer(notebook)
+    history_query = """
+Generate a list of strings from the provided notebook where each string is a summary of an event. 
+An event is either a single cell OR a parent query cell + a child code cell.
```

Comment:
A cell isn't an event, it's just something that exists.
The language implies that there should be an inherent order, but it's unclear which of the two possible orders that one could infer should be used.
The first order is the order of the cells in the notebook. But the cells are not events. Cells can just have information, or they could be code that was not run, or they could be code that did run and produced an output.
The second order is the execution order of code cells based on the execution_count.

The events that I see in these notebooks are:
- Agent Query
  - A question is asked and an answer is given (and the answer may or may not be code that place in a code block)
- Code cell execution
  - Code is executed and 'outputs'  is populated with the outputs and results.

I'm thinking we probably want to word it something that takes both in to account.

Generated list ... where each entry is a summary of a cell in the notebook in order from top to bottom. If the cell is a code-cell, summarize the code and the result, along with what order it was executed in based on the execute count. If the cell is a markdown cell with metadata where beaker_cell_type = "query", then this cell contains an interaction between the user and the LLM agent. The Agent may have provided a "response" message or may have responded with code that is added in subsequent cell. For these query cells, make sure to summarize the query and the response, if provided, and note if and which code cell was generated, if a code cell was generated by the query.

I think this, if it returns how I'm imagining, will provide more details for the summarization step below.

Reply:
I incorporated this into the prompt. If this looks good to you, feel free to resolve.